google/cloud/forseti/scanner/audit/external_project_access_rules_engine.py

```diff
@@ -0,0 +1,286 @@
+import re
+
+from google.cloud.forseti.common.gcp_type import resource as resource_mod
+from google.cloud.forseti.common.gcp_type import resource_util
+from google.cloud.forseti.common.util import logger
+from google.cloud.forseti.scanner.audit import errors as audit_errors
+from google.cloud.forseti.scanner.audit import base_rules_engine as bre
+
+LOGGER = logger.get_logger(__name__)
+
+
+class ExternalProjectAccessRulesEngine(bre.BaseRulesEngine):
+    """Rules engine for External Project Access."""
+
+    def __init__(self, rules_file_path, snapshot_timestamp=None):
+        """Initialize.
+
+        Args:
+            rules_file_path (str): file location of rules
+            snapshot_timestamp (str): snapshot timestamp. Defaults to None.
+                If set, this will be the snapshot timestamp
+                used in the engine.
+        """
+        super(ExternalProjectAccessRulesEngine,
+              self).__init__(rules_file_path=rules_file_path)
+        self.rule_book = None
+
+    def build_rule_book(self, global_configs=None):
+        """Build ExternalProjectAccess rule book from the rules definition file.
+
+        Args:
+            global_configs (dict): Inventory configurations.
+        """
+        self.rule_book = (
+            ExternalProjectAccessRuleBook(
+                global_configs,
+                self._load_rule_definitions()))
+
+    # TODO: The naming is confusing and needs to be fixed in all scanners.
+    def find_policy_violations(self, user_email,
+                               project_ancestry,
+                               force_rebuild=False):
+        """Determine whether project ancestry violates rules.
+
+        Args:
+            user_email (str): The user's e-mail
+            project_ancestry (list): List of ancestries which turn
+                                     out to a list of resources.
+            force_rebuild (bool): Force the rebuild of the rule book
+
+        Returns:
+             generator: A generator of rule violations.
+        """
+
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+
+        violations = itertools.chain(
+            self.rule_book.find_policy_violations(
+                user_email, project_ancestry)
+        )
+
+        return violations
+
+
+class ExternalProjectAccessRuleBook(bre.BaseRuleBook):
+    """The RuleBook for ExternalProjectAccess resources."""
+
+    # Class variable for matching the ancestor during rule validation
+    ancestor_pattern = re.compile(r'^organizations/\d+$|^folders/\d+$')
+
+    def __init__(self, global_configs=None, rule_defs=None):
+        """Initialization.
+
+        Args:
+            rule_defs (dict): rule definitons
+            global_configs (dict): Inventory configurations.
+        """
+        super(ExternalProjectAccessRuleBook, self).__init__()
+        self.resource_rules_map = dict()
+
+        # Create a default rule for the current organization as
```

Comment:
My desire is to have this require as little pre-configuration as possible.  I've commented out the lines in the YAML file and indicated the rule is an example.

Reply:
Hmmm, I think that we should not embed a default rule in the code.  It would give a lot more visibility & flexibility to have the default rule to be placed and used in the rules.yaml file.  It also maintains consistency with the rest our scanner rule files.

If you are concerned about pre-configuration, there is a solution.  Our installer can look for a specific placeholder, in this case `{ORGANIZATION_ID}`, and populate it with the right values from the environment.
Would this work for you?
```
rules:
- name: Find external project access in organization.
  allowed_ancestor: organizations/{ORGANIZATION_ID}
```

Reply:
It sure does!